### PR TITLE
core[minor]: change how tool error handling works

### PR DESCRIPTION
**Pull request by isahers1**
- error handling can now be passed through decorator
- all errors thrown will be ValidationErrors (tool call was formatted incorrectly) or ToolExceptions (any error that occurs while running the tool)

### CODE REVIEW

File: libs/core/langchain_core/tools/base.py [hunk 1]
```diff
@@ -649,43 +649,50 @@ def run(
...
-            else:
-                content = response
-            status = "success"
-        except ValidationError as e:
-            if not self.handle_validation_error:
+                status = "error"
+            except ToolException as e:
+                raise e
+            except Exception as e:
+                raise ToolException(str(e)) from e
```
Comment (efriis) [inline libs/core/langchain_core/tools/base.py:686]: do we want to format this any differently? I believe this has impact on agent performance

File: libs/core/langchain_core/tools/base.py [hunk 2]
```diff
@@ -649,43 +649,50 @@ def run(
...
+            except KeyboardInterrupt as e:
```
Comment (efriis) [inline libs/core/langchain_core/tools/base.py:687]: does this not log as an exception? Will this ever get hit after `except Exception`?

**Review by efriis — COMMENTED** [review]
Believe this will technically be breaking for anyone doing error handling on their tools currently - could also be viewed as a fix. Will let @baskaryan comment

File: libs/core/langchain_core/tools/base.py (hunk 2 above)
Comment (isahers1) [reply libs/core/langchain_core/tools/base.py:687]: keyboard interrupt is treated differently at least according my basic testing with this tool:
```@tool
def my_tool(x: int):
    """My tool"""
    time.sleep(100)
    return x*5```
    
I was able to run and interrupt as normal

File: libs/core/langchain_core/tools/base.py (hunk 1 above)
Comment (isahers1) [reply libs/core/langchain_core/tools/base.py:686]: I don't understand why this impacts performance - but happy to reformat to whatever just lmk what you had in mind

File: libs/core/langchain_core/tools/convert.py [hunk 3]
```diff
@@ -42,6 +48,14 @@ def tool(
         error_on_invalid_docstring: if ``parse_docstring`` is provided, configure
             whether to raise ValueError on invalid Google Style docstrings.
             Defaults to True.
+        handle_tool_error: Handle the content of the ToolException thrown. If False
+              do nothing, If True returns 'Tool execution error'. If string then
+              returns that string directly, if callable converts error to string to be returned.
+              Defaults to False.
+          handle_validation_error: Handle the content of the ValidationError thrown. If False
```
Comment (vbarda) [inline libs/core/langchain_core/tools/convert.py:55]: nit - docstring misaligned

Comment (vbarda) [reply libs/core/langchain_core/tools/convert.py:55]: also, i think the param description is a bit confusing (which is true for both the original description in the `BaseTool` class as well) - this is not handling the content of the error, but content of `ToolMessage` returned if the error is thrown

**Pull request closed without merging**